doc/developer/design/20231213_compute_operator_frontier_tracking.md

```diff
@@ -0,0 +1,187 @@
+
+1. [Assigning IDs to LIR nodes during query planning](#lir-node-ids)
+2. [Creating a separate region for each LIR node during Compute rendering](#lir-dataflow-regions)
+3. [Collecting LIR region frontier information during dataflow processing](#frontier-logging)
+4. [Reporting LIR region frontier information to the controller](#reporting-introspection-data)
+
+While there is some amount of lifting required to implement all these pieces, our hope is that they will prove generally useful.
+(1) and (2) together allow mapping low-level dataflow information back to `EXPLAIN` plans and will be useful for making our internal memory visualizer user-facing.
+(4) gives us a reliable way to report dataflow introspection information in SQL, which will be useful in alleviating some of the pain currently caused by the bad ergonomics of our per-replica introspection mechanism.
+
+### LIR Node IDs
+
+To make LIR nodes identifiable, we assign IDs to them (referred to as _LIDs_ in the following).
+LIDs need to be unique only within a dataflow, as they can be namespaced by the dataflow ID, or by the ID of the SQL object that created the dataflow.
+
+This design proposes that LIDs should be increasing integers.
+This is the simplest ID format and there is currently no need for anything more complex.
+Since both the `EXPLAIN` output and the introspection tables that will expose operator frontier information are part of Materialize's unstable API, we are free to change the ID format later, should the need arise.
+
+LIDs need to be assigned as part of query planning.
+Because LIR-level optimizations can change the plan by removing and adding nodes, ID assignment should ideally happen after all optimization have been performed.
+At the same time, we need to make sure that LIDs are assigned in the plans produced by `EXPLAIN PHYSICAL PLAN`, so we can show them in the explain output.
+To satisfy both requirements, we propose performing LID assignment at the end of `finalize_dataflow`.
+
+Alternatively, we could assign an ID immediately upon creation of each LIR node.
+When later optimizations add/remove LIR nodes, this would lead to IDs that are seemingly missing or misordered.
+To avoid potential confusion, we reject this approach.
+
+As indicated above, we need to adjust the explain code to show LIDs in the output of `EXPLAIN PHYSICAL PLAN`.
+To reduce the noise level, inclusion of LIDs will be controlled by an output modifier, and disabled by default.
+
+### LIR Dataflow Regions
+
+To be able map LIR nodes to dataflow operators, Compute will render a separate dataflow region for each LIR node and associate it with the respective LID.
+
+The ID association is performed by adding the LID to the operator name.
+Exposing the mapping between operator IDs and LIDs in a SQL table is something we can consider as an alternative, but it seems to introduce unnecessary complexity at this point.
+
+### Frontier Logging
```

Comment:
I'm trying to understand what logging means. Is it:
1. Updating the values for each dataflow in `mz_internal.mz_dataflow_operator_frontiers_per_worker`?
2. Appending a new row in `mz_internal.mz_dataflow_operator_frontiers_per_worker`?

As far as I understand, it is option 1, and it is called logging because you can review the history using the `AS OF` clause. Is that correct?

Reply:
Yeah, it's the former. It's called logging because its produces log events, not a persistent log. The log events are then processed in some way, in this case they are reported to the controller, which writes them to a persist collection with upsert semantics. So no append-only log is created.